core/src/com/unciv/logic/automation/WorkerAutomation.kt

```diff
@@ -333,15 +333,21 @@ class WorkerAutomation(
     private fun chooseImprovement(unit: MapUnit, tile: TileInfo): TileImprovement? {
 
         // turnsToBuild is what defines them as buildable
-        val tileImprovements = ruleSet.tileImprovements.filter {
-            it.value.turnsToBuild != 0 && tile.canImprovementBeBuiltHere(it.value, tile.hasViewableResource(civInfo)) }
-        val uniqueImprovement = tileImprovements.values
-            .firstOrNull { it.uniqueTo == civInfo.civName }
+        val potentialTileImprovements = ruleSet.tileImprovements.filter {
+            unit.canBuildImprovement(it.value, tile)
+                    && tile.canImprovementBeBuiltHere(it.value, tile.hasViewableResource(civInfo))
+                    && (it.value.uniqueTo == null || it.value.uniqueTo == unit.civInfo.civName)
```

Comment:
Would it be better for performance to do this check first?

Reply:
Perhaps, I considered this an edge case but you're right